core/executor/src/wasm_runtimes_cache.rs

```
@@ -0,0 +1,328 @@
+		ext: &mut E,
+		default_heap_pages: Option<u64>,
+	) -> Result<Rc<CachedRuntime>, Error> {
+		let code_hash = ext
+			.original_storage_hash(well_known_keys::CODE)
+			.ok_or(Error::InvalidCode)?;
+
+		// This is direct result from fighting with borrowck.
+		let handle_result =
+			|cached_result: &Result<Rc<CachedRuntime>, CacheError>| match *cached_result {
+				Err(_) => Err(Error::InvalidCode),
+				Ok(ref cached_runtime) => Ok(Rc::clone(cached_runtime)),
+			};
+
+		match self.instances.entry(code_hash.into()) {
+			Entry::Occupied(o) => handle_result(o.get()),
+			Entry::Vacant(v) => {
+				trace!(target: "runtimes_cache", "no instance found in cache, creating now.");
+				let result = Self::create_wasm_instance(wasm_executor, ext, default_heap_pages);
+				if let Err(ref err) = result {
+					warn!(target: "runtimes_cache", "cannot create a runtime: {:?}", err);
+				}
+				handle_result(v.insert(result))
+			}
+		}
+	}
+
+	fn create_wasm_instance<E: Externalities<Blake2Hasher>>(
+		wasm_executor: &WasmExecutor,
+		ext: &mut E,
+		default_heap_pages: Option<u64>,
+	) -> Result<Rc<CachedRuntime>, CacheError> {
+		let code = ext
+			.original_storage(well_known_keys::CODE)
+			.ok_or(CacheError::CodeNotFound)?;
+		let module = WasmModule::from_buffer(&code).map_err(|_| CacheError::InvalidModule)?;
+
+		// Extract the data segments from the wasm code.
+		//
+		// A return of this error actually indicates that there is a problem in logic, since
+		// we just loaded and validated the `module` above.
+		let data_segments = extract_data_segments(&code).ok_or(CacheError::CantDeserializeWasm)?;
+
+		let heap_pages = ext
+			.storage(well_known_keys::HEAP_PAGES)
+			.and_then(|pages| u64::decode(&mut &pages[..]))
+			.or(default_heap_pages)
+			.unwrap_or(DEFAULT_HEAP_PAGES);
+
+		// Instantiate this module.
+		let instance = WasmExecutor::instantiate_module::<E>(heap_pages as usize, &module)
+			.map_err(|_| CacheError::Instantiation)?;
```

This error shouldn't get silently dropped -- it made it harder to find the issue with `__heap_base`.